lib/ClangImporter/ImportType.cpp

```diff
@@ -2647,43 +2647,46 @@ ArgumentAttrs ClangImporter::Implementation::inferDefaultArgument(
   } else if (const clang::TypedefType *typedefType =
                  type->getAs<clang::TypedefType>()) {
     // Get the AvailabilityAttr that would be set from CF/NS_OPTIONS
-    if (importer::isUnavailableInSwift(typedefType->getDecl(), nullptr, true)) {
+    clang::TypedefNameDecl *typedefDecl = typedefType->getDecl();
+    if (importer::isUnavailableInSwift(typedefDecl, nullptr, true)) {
       // If we've taken this branch it means we have an enum type, and it is
       // likely an integer or NSInteger that is being used by NS/CF_OPTIONS to
       // behave like a C enum in the presence of C++.
-      auto enumName = typedefType->getDecl()->getName();
-      ArgumentAttrs argumentAttrs(DefaultArgumentKind::None, true, enumName);
-      auto camelCaseWords = camel_case::getWords(enumName);
-      for (auto it = camelCaseWords.rbegin(); it != camelCaseWords.rend();
-           ++it) {
-        auto word = *it;
-        auto next = std::next(it);
-        if (camel_case::sameWordIgnoreFirstCase(word, "options")) {
-          argumentAttrs.argumentKind = DefaultArgumentKind::EmptyArray;
-          return argumentAttrs;
+      auto enumName = typedefDecl->getName();
+      // Find the next decl in the same context. If this typedef is a part of an
+      // NS/CF_OPTIONS declaration, the next decl will be an enum.
+      auto declsInContext = typedefDecl->getDeclContext()->decls();
+      auto declIter = llvm::find(declsInContext, typedefDecl);
+      if (declIter != declsInContext.end())
+        declIter++;
+      if (declIter != declsInContext.end()) {
+        if (auto enumDecl = dyn_cast<clang::EnumDecl>(*declIter)) {
+          // If this is a NS/CF_OPTIONS declaration, the enum will have no name,
+          // and all enum cases will likely be prefixed with the name of the
+          // enum.
+          bool looksLikeCFOptions =
+              !enumDecl->getIdentifier() &&
+              enumDecl->hasAttr<clang::FlagEnumAttr>() &&
+              enumDecl->hasAttr<clang::EnumExtensibilityAttr>();
+          if (looksLikeCFOptions) {
+            for (auto caseDecl : enumDecl->enumerators()) {
+              if (!caseDecl->getName().starts_with(enumName)) {
+                looksLikeCFOptions = false;
+                break;
+              }
+            }
+          }
+          if (looksLikeCFOptions) {
+            ArgumentAttrs argumentAttrs(DefaultArgumentKind::None, true,
+                                        enumName);
+            for (auto word : llvm::reverse(camel_case::getWords(enumName))) {
+              if (camel_case::sameWordIgnoreFirstCase(word, "options")) {
+                argumentAttrs.argumentKind = DefaultArgumentKind::EmptyArray;
+              }
+            }
+            return argumentAttrs;
+          }
         }
-        if (camel_case::sameWordIgnoreFirstCase(word, "units"))
```

Comment:
We're relaxing the requirements imposed on the name of the NS_OPTIONS type. Previously we were only detecting type names ending with certain words here, now we're ignoring the type name (since custom frameworks can declare NS_OPTIONS types with arbitrary names) and relying on the AST structure of an NS_OPTIONS expansion.
So the type name detection cases were intentionally removed.